proposals/NNNN-library-evolution.md

```diff
@@ -0,0 +1,369 @@
+
+- **It is not guaranteed to be "trivial"** ([in the C++ sense][trivial]). A frozen struct containing a class reference or closure still requires reference-counting when copied and when it goes out of scope.
+
+- **It does not necessarily have a known size or alignment.** A generic frozen struct's layout might depend on the generic argument provided at run time.
+
+- **Even concrete instantiations may not have a known size or alignment.** A frozen struct with a field that's a *non-*frozen, has a size that may not be known until run time.
+
+- **It is not guaranteed to use the same layout as a C struct with a similar "shape".** If such a struct is necessary, it should be defined in a C header and imported into Swift.
+
+- **The fields are not guaranteed to be laid out in declaration order.** The compiler may choose to reorder fields, for example to minimize padding while satisfying alignment requirements.
+
+That said, the compiler is allowed to use its knowledge of the struct's contents and layout to derive any of these properties. For instance, the compiler can statically prove that copying `Point` is "trivial", because each of its members has a statically-known type that is "trivial". However, depending on this at the language level is not supported, with two exceptions:
+
+1. The run-time memory layout of a struct with a single field is always identical to the layout of the instance property on its own, whether the struct is declared `@frozen` or not. This has been true since Swift 1. (This does not extend to the calling convention, however. If the struct is not frozen, it will be passed indirectly even if its single field is frozen and thus _can_ be passed directly)
+
+2. The representation of `nil` for any type that is "nullable" in C / Objective-C is the same as the representation of `nil` or `NULL` in those languages. This includes class references, class-bound protocol references, class type references, unsafe-pointers, `@convention(c)` functions, `@convention(block)` functions, OpaquePointer, Selector, and NSZone. This has been true since Swift 3 ([SE-0055][]).
+
+This proposal does not change either of these guarantees.
+
+  [trivial]: https://docs.microsoft.com/en-us/cpp/cpp/trivial-standard-layout-and-pod-types
+  [SE-0055]: https://github.com/apple/swift-evolution/blob/master/proposals/0055-optional-unsafe-pointers.md
+
+### `@frozen` on `enum` types
+
+Marking an enum as `@frozen` will similarly allow the compiler to optimize away runtime calls.
+
+In addition, marking an enum as frozen restores the ability of a library user to exhaustively switch over that enum without an `@unknown default:`, because it guarantees no further cases will be added.
+
+Once frozen, all changes made to an enum's cases affect its ABI.
+
+## Naming
+
+`@frozen` was used here, as originally used in [SE-0192][], for both fixed-layout structs and enums. 
```

Comment:
Please fill in these links and make any other editorial changes you intend to make before this goes up.

Reply:
The link's there, it's (implicit) reference style markdown link. Renders correctly in GitHub.